feat/docs(editorconfig): add support for spelling_language

marcg03 opened the pull request:
This fixes #28626.

For the other **editorconfig** properties it's easier to check for validity since they are either a number or part of an enum with few elements. The problem for `spelling_language` is that it is part of a greater enum since it's a concatenation of a [ISO 639 language code](https://www.loc.gov/standards/iso639-2/php/code_list.php) and an optional [ISO 3166 territory identifier](https://www.iso.org/obp/ui/#search).

For now the code only checks that the value provided in `spelling_language` is of the format [a-z][a-z] or [a-z][a-z]-[a-z][a-z] (regex), but it doesn't check whether it is a valid ISO639/ISO3166 code and thus when inputting an invalid value the `spelllang` option is also set to an unknown language (this results in a prompt asking you to download the unknown language and then failing and continuing if accepted).

Commit:
feat(editorconfig): add spelling_language support

Commit:
docs(editorconfig): generate spelling_language docs

gpanders commented on runtime/lua/editorconfig.lua:
```diff
@@ -191,6 +191,27 @@ function properties.insert_final_newline(bufnr, val)
+--- A code of the format ss or ss-TT, where ss is an ISO 639 language code and TT is an ISO 3166 territory identifier.
+--- Sets the 'spell' and 'spelllang' options.
+function properties.spelling_language(bufnr, val)
+  local errorMsg =
```
Nit: use snake case

```suggestion
  local error_msg =
```

gpanders commented on runtime/lua/editorconfig.lua:
```diff
@@ -191,6 +191,27 @@ function properties.insert_final_newline(bufnr, val)
+
+  assert(val:len() == 2 or val:len() == 5, errorMsg)
+
+  local languageCode = val:sub(1, 2):lower()
```
```suggestion
  local language_code = val:sub(1, 2):lower()
```

gpanders commented on runtime/lua/editorconfig.lua:
```diff
@@ -191,6 +191,27 @@ function properties.insert_final_newline(bufnr, val)
+  if val:len() == 2 then
+    vim.bo[bufnr].spelllang = languageCode
+  else
+    local territoryCode = val:sub(4, 5):lower()
```
```suggestion
    local territory_code = val:sub(4, 5):lower()
```

gpanders reviewed (commented):
Looks good on first glance, thanks! Will give it a test whenever you mark it ready for review.

Commit:
style: use snake_case not camelCase

gpanders commented:
Can you please also update `news.txt` and add a test case to `editorconfig_spec.lua`?

Commit:
test(editorconfig): check spelling_language support

Commit:
chore: update news.txt

sus-domesticus commented:
I encountered problems with the `test_case` method (in `editorconfig_spec.lua`).
The error is from passing `{ buf = 0 }` to `nvim_get_option_value` when getting the value of `spell` because it's **window-local** not **buffer-local**.
![error](https://github.com/neovim/neovim/assets/134197728/5c5a29a7-66a4-41d8-a747-0ac54f5c67f4)

Should I adapt `test_case` for the **window-local** `spell` option?


https://github.com/sus-domesticus/neovim/blob/8372a48a3564564577ca5096a0c4584f0cdcd9a8/test/functional/plugin/editorconfig_spec.lua#L14-L22

Commit:
test(editorconfig): handle opts with different scopes

The `test_case` method presumed that all options set by editorconfig
were *buffer-local* so it called `nvim_get_option_value` with `{ buf = 0
}` even for *window-local* options such as `spell`.

Now `nvim_get_option_info2` is used to get the scope of the option. This
scope can be: "global", "win" or "buf".

zeertzjq commented on test/functional/plugin/editorconfig_spec.lua:
```diff
@@ -16,8 +16,18 @@ local testdir = 'Xtest-editorconfig'
+    local opt_info = api.nvim_get_option_info2(opt)
+    if opt_info.scope == 'win' then
+      eq(val, api.nvim_get_option_value(opt, { win = 0 }), name)
+    else
+      if opt_info.scope == 'buf' then
```
This can use `elseif`

Commit:
test(editorconfig): use elseif not else and if

sus-domesticus commented:
Looks like I forgot to add some parameters. It's weird because when I ran `make test` locally I didn't get an error.

Commit:
test(editorconfig): add missing params

zeertzjq commented on test/functional/plugin/editorconfig_spec.lua:
```diff
@@ -16,8 +16,16 @@ local testdir = 'Xtest-editorconfig'
+
   for opt, val in pairs(expected) do
-    eq(val, api.nvim_get_option_value(opt, { buf = 0 }), name)
+    local opt_info = api.nvim_get_option_info2(opt)
```
```suggestion
    local opt_info = api.nvim_get_option_info2(opt, {})
```

sus-domesticus commented on runtime/lua/editorconfig.lua:
```diff
@@ -191,6 +191,27 @@ function properties.insert_final_newline(bufnr, val)
+
+  local language_code = val:sub(1, 2):lower()
+  assert(language_code:match('%l%l'), error_msg)
+  if val:len() == 2 then
+    vim.bo[bufnr].spelllang = language_code
+  else
+    local territory_code = val:sub(4, 5):lower()
+    assert(language_code:match('%l%l'), error_msg)
+    vim.bo[bufnr].spelllang = language_code .. '_' .. territory_code
```
Yes you are right! And also on line **208** instead of checking `territory_code` I am checking `language_code` again so I'll fix that as well.

Commit:
test(editorconfig): check spelling_language separator

A *GOOD* example of spelling_language is en-US. ('-' must be used as a
separator)

A *BAD* example of spelling_language is en/US or en_US.

Commit:
test(editorconfig): check territory_code format

justinmk commented on test/functional/plugin/editorconfig_spec.lua:
```diff
@@ -93,6 +101,12 @@ setup(function()
+    spelling_language = en
+
+    [long_spelling_language.txt]
+    spelling_language = en-US
```
probably should test something other than the default

Commit:
test(editorconfig): change spelling_language properties

These were changed to use something other than the default values.

gpanders reviewed (approved):
Sorry for the delay. Just needs a rebase onto the new `news.txt` file and then should be good.

gpanders commented on runtime/doc/news.txt:
```diff
@@ -130,6 +130,9 @@ PLUGINS
 
 • TODO
 
+• Editorconfig
```
Nit
```suggestion
• EditorConfig
```

Can also delete the `TODO` just above

gpanders commented on runtime/lua/editorconfig.lua:
```diff
@@ -190,6 +190,29 @@ function properties.insert_final_newline(bufnr, val)
 end
 
+--- A code of the format ss or ss-TT, where ss is an ISO 639 language code and TT is an ISO 3166 territory identifier.
+--- Sets the 'spell' and 'spelllang' options.
```
I am not sure how I feel about this also setting `'spell'`. Someone may want to apply `spelling_language` globally (using a `[*]` glob) so that _when they do_ enable spell checking, it will use the specified language. But that does not imply they necessarily want spell checking always enabled.

I think we should remove `'spell'` for now and set only `'spelllang'`. This is reversible if we need to re-enable it later.

sus-domesticus replied:
The spec also doesn't say anything about enabling spell checking.
> Sets the natural language that should be used...

Commit:
fix(news): remove placeholder TODO; fix name

Commit:
fix(editorconfig): don't set the spell option

A spelling_language key-value pair in an editorconfig-file should only
set `spelllang` and leave `spell` as is.

Commit:
test(editorconfig): adjust spelling_language tests

The `spell` option shouldn't be checked by spelling_language-tests since
spelling_language should only change the `spelllang` option.

Commit:
docs(editorconfig): update spelling_language description

sus-domesticus commented:
I could also try to tidy the commit history a bit if you think it's needed.

gpanders commented:
Thanks!
